File: Migration/fileSystemUtils.js
const fs = require('fs');
const path = require('path');

/**
 * Utility class for file system operations related to project analysis.
 */
class FileSystemUtils {
    constructor(baseDir, excludePatterns = []) {
        this.baseDir = baseDir;
        this.excludePatterns = excludePatterns;
        this.allFiles = [];
        this.fileContents = new Map();
        this.projectStructure = {
            directories: new Set(),
            filesByType: {},
            filesByDir: {},
            dirCategories: {
                api: new Set(),
                models: new Set(),
                ui: new Set(),
                tests: new Set(),
                services: new Set()
            }
        };
        this.keywordIndex = new Map(); // Index for keywords found in files
    }

    /**
     * Discovers all relevant files in the base directory, excluding specified patterns.
     * Populates project structure information.
     */
    discoverFiles() {
        console.log("Discovering files...");
        this.allFiles = this._walkDir(this.baseDir);
        console.log(`Found ${this.allFiles.length} files`);
        return this.allFiles;
    }

    _walkDir(dir, filelist = []) {
        try {
            const files = fs.readdirSync(dir);

            for (const file of files) {
                const filepath = path.join(dir, file);
                const stat = fs.statSync(filepath);
                const relativePath = path.relative(this.baseDir, filepath);

                // Skip if matches any exclude pattern
                if (this.excludePatterns.some(pattern => pattern.test(relativePath))) {
                    continue;
                }

                if (stat.isDirectory()) {
                    // Track directory for project structure analysis
                    if (!relativePath.startsWith('.')) { // Avoid hidden dirs like .git
                        this.projectStructure.directories.add(relativePath || '.');
                        this._categorizeDirectory(relativePath);
                    }
                    filelist = this._walkDir(filepath, filelist);
                } else {
                    // Process file
                    const ext = path.extname(filepath).toLowerCase();

                    // Track by file type
                    if (!this.projectStructure.filesByType[ext]) {
                        this.projectStructure.filesByType[ext] = [];
                    }
                    this.projectStructure.filesByType[ext].push(relativePath);

                    // Track by directory
                    const dirName = path.dirname(relativePath);
                    if (!this.projectStructure.filesByDir[dirName]) {
                        this.projectStructure.filesByDir[dirName] = [];
                    }
                    this.projectStructure.filesByDir[dirName].push(relativePath);

                    filelist.push(filepath);
                }
            }

            return filelist;
        } catch (err) {
            console.error(`Error reading directory ${dir}:`, err.message);
            return filelist;
        }
    }

    _categorizeDirectory(relativePath) {
        const categories = this.projectStructure.dirCategories;
        if (relativePath.includes('api') || relativePath.includes('routes')) {
            categories.api.add(relativePath);
        } else if (relativePath.includes('model') || relativePath.includes('entity')) {
            categories.models.add(relativePath);
        } else if (relativePath.includes('component') || relativePath.includes('ui') || relativePath.includes('pages') || relativePath.includes('features')) {
             // Added pages/features
            categories.ui.add(relativePath);
        } else if (relativePath.includes('test') || relativePath.includes('spec')) {
            categories.tests.add(relativePath);
        } else if (relativePath.includes('service') || relativePath.includes('util') || relativePath.includes('core') || relativePath.includes('shared')) {
             // Added core/shared
            categories.services.add(relativePath);
        }
    }

    /**
     * Reads the content of discovered text files, skipping binaries and large files.
     * Populates the fileContents map and keyword index.
     */
    readFileContents() {
        console.log("Reading file contents with advanced processing...");

        const binarySignatures = [ /* ... signatures ... */ ]; // Keep signatures as before
        const skipExtensions = new Set([ /* ... extensions ... */ ]); // Keep skip extensions
        const textExtensions = new Set([ /* ... extensions ... */ ]); // Keep text extensions

        const fileStats = { read: 0, skipped: 0, binary: 0, tooLarge: 0, error: 0, emptyFile: 0 };

        for (const filePath of this.allFiles) {
            try {
                const ext = path.extname(filePath).toLowerCase();
                if (skipExtensions.has(ext)) {
                    fileStats.skipped++;
                    continue;
                }

                const stats = fs.statSync(filePath);
                if (stats.size === 0) {
                    fileStats.emptyFile++;
                    continue;
                }

                const isKnownText = textExtensions.has(ext);
                const sizeLimit = isKnownText ? 5 * 1024 * 1024 : 1024 * 1024;
                if (stats.size > sizeLimit) {
                    fileStats.tooLarge++;
                    continue;
                }

                if (!isKnownText && this._isLikelyBinary(filePath, binarySignatures)) {
                    fileStats.binary++;
                    continue;
                }

                const content = fs.readFileSync(filePath, 'utf8');
                this.fileContents.set(filePath, content);
                this._indexFileKeywords(filePath, content); // Index keywords
                fileStats.read++;

            } catch (err) {
                fileStats.error++;
                console.error(`Error reading ${filePath}:`, err.message);
            }
        }

        console.log(`Read ${fileStats.read} files, ${fileStats.skipped + fileStats.binary + fileStats.tooLarge + fileStats.emptyFile} skipped`);
        console.log(`  Skipped: ${fileStats.skipped} by extension, ${fileStats.binary} binary, ${fileStats.tooLarge} too large, ${fileStats.emptyFile} empty`);
        return this.fileContents;
    }

    _isLikelyBinary(filePath, binarySignatures) {
        try {
            const fd = fs.openSync(filePath, 'r');
            const buffer = Buffer.alloc(8);
            fs.readSync(fd, buffer, 0, 8, 0);
            fs.closeSync(fd);

            const isBinarySig = binarySignatures.some(signature => {
                for (let i = 0; i < signature.length; i++) {
                    if (buffer[i] !== signature[i]) return false;
                }
                return true;
            });

            const hasNullByte = buffer.includes(0x00);
            return isBinarySig || hasNullByte;
        } catch (err) {
            console.error(`Error checking binary status for ${filePath}:`, err.message);
            return true; // Assume binary on error
        }
    }

     /**
     * Index file keywords for faster searching.
     * (Internal helper method)
     */
    _indexFileKeywords(filePath, content) {
        // List of important keywords to index
        const keywords = [
            // Models
            'struct', 'enum', 'trait', 'impl', 'class', 'interface', 'type', 'model', 'entity',
            // API
            'fn', 'function', 'route', 'get', 'post', 'put', 'delete', 'api', 'endpoint', 'handler', 'router', 'controller',
            // UI
            'component', 'function', 'render', 'return', 'useState', 'useEffect', 'props', 'view', 'page', 'layout',
            // Tests
            'test', 'describe', 'it', 'expect', 'assert', 'mock',
            // General
            'use', 'mod', 'import', 'require', 'export', 'async', 'await', 'pub', 'static', 'const'
        ];

        // Check for each keyword
        for (const keyword of keywords) {
            if (content.includes(keyword)) {
                if (!this.keywordIndex.has(keyword)) {
                    this.keywordIndex.set(keyword, new Set());
                }
                this.keywordIndex.get(keyword).add(filePath);
            }
        }
    }


    /**
     * Finds files matching a list of regex patterns.
     */
    findFilesByPatterns(patterns) {
        return this.allFiles.filter(filePath => {
            const relativePath = path.relative(this.baseDir, filePath);
            return patterns.some(pattern => pattern.test(relativePath));
        });
    }

    /**
     * Gets statistics about a directory (file count, total size).
     */
    getDirectoryStats(dirPath) {
        let fileCount = 0;
        let totalSize = 0;
        const fullPath = path.join(this.baseDir, dirPath);

        const walk = (dir) => {
            try {
                const files = fs.readdirSync(dir);
                for (const file of files) {
                    const filepath = path.join(dir, file);
                    const stat = fs.statSync(filepath);
                    const relativePath = path.relative(this.baseDir, filepath);

                    if (this.excludePatterns.some(pattern => pattern.test(relativePath))) {
                        continue;
                    }

                    if (stat.isDirectory()) {
                        walk(filepath);
                    } else {
                        fileCount++;
                        totalSize += stat.size;
                    }
                }
            } catch (err) {
                console.error(`Error getting stats for directory ${dir}:`, err.message);
            }
        };

        walk(fullPath);
        return { fileCount, totalSize };
    }

    // --- Getters for accessed properties ---
    getAllFiles() {
        return this.allFiles;
    }

    getFileContentsMap() {
        return this.fileContents;
    }

    getProjectStructure() {
        return this.projectStructure;
    }

     getKeywordIndex() {
        return this.keywordIndex;
    }

    /**
     * Determines the category of a directory based on its path.
     * @param {string} dirPath - The directory path to categorize
     * @return {string} The category name (models, api, ui, tests, etc.)
     */
    getDirCategory(dirPath) {
        // Normalize path for consistent pattern matching
        const normalizedPath = dirPath.replace(/\\/g, '/').toLowerCase();
        
        if (normalizedPath.includes('/models') || normalizedPath.includes('/entities')) {
          return 'models';
        } else if (normalizedPath.includes('/api') || normalizedPath.includes('/routes')) {
          return 'api';
        } else if (normalizedPath.includes('/components') || normalizedPath.includes('/views') || normalizedPath.includes('/pages')) {
          return 'ui';
        } else if (normalizedPath.includes('/tests') || normalizedPath.includes('/__tests__') || normalizedPath.includes('/test')) {
          return 'tests';
        } else if (normalizedPath.includes('/services') || normalizedPath.includes('/lib')) {
          return 'service';
        } else if (normalizedPath.includes('/utils') || normalizedPath.includes('/helpers')) {
          return 'utility';
        } else if (normalizedPath.includes('/docs') || normalizedPath.includes('/documentation')) {
          return 'documentation';
        } else if (normalizedPath.includes('/config')) {
          return 'configuration';
        }
        
        return 'other';
    }

    /**
     * Filter files by pattern
     * @param {RegExp} pattern - The regex pattern to match file paths
     * @returns {string[]} - Filtered file paths
     */
    filterFiles(pattern) {
        if (!this.files) return [];
        return Object.keys(this.files)
            .filter(file => pattern.test(file));
    }

    /**
     * Get file statistics
     * @returns {Object} File statistics with counts by file type
     */
    getFileStats() {
        const allFiles = this.getAllFiles();
        
        // Count files by type
        const js = allFiles.filter(file => /\.(js|jsx|ts|tsx)$/.test(file)).length;
        const rust = allFiles.filter(file => /\.rs$/.test(file)).length;
        const other = allFiles.length - js - rust;
        
        return {
          total: allFiles.length,
          js,
          rust,
          other
        };
    }
}

// Re-add the binary signatures and skip/text extensions here
FileSystemUtils.prototype._isLikelyBinary = function(filePath, binarySignatures = []) {
    // Binary file signatures/magic numbers (hex) - Moved here for encapsulation
    const defaultBinarySignatures = [
      [0xFF, 0xD8], // JPEG
      [0x89, 0x50, 0x4E, 0x47], // PNG
      [0x47, 0x49, 0x46], // GIF
      [0x50, 0x4B, 0x03, 0x04], // ZIP/JAR/DOCX
      [0x25, 0x50, 0x44, 0x46], // PDF
    ];
    const signaturesToCheck = binarySignatures.length > 0 ? binarySignatures : defaultBinarySignatures;

    try {
        const fd = fs.openSync(filePath, 'r');
        const buffer = Buffer.alloc(8); // Read first 8 bytes
        fs.readSync(fd, buffer, 0, 8, 0);
        fs.closeSync(fd);

        const isBinarySig = signaturesToCheck.some(signature => {
            for (let i = 0; i < signature.length; i++) {
                if (buffer[i] !== signature[i]) return false;
            }
            return true;
        });

        const hasNullByte = buffer.includes(0x00);
        return isBinarySig || hasNullByte;
    } catch (err) {
        console.error(`Error checking binary status for ${filePath}:`, err.message);
        return true; // Assume binary on error
    }
};

FileSystemUtils.prototype.readFileContents = function() {
    console.log("Reading file contents with advanced processing...");

    // Moved here for encapsulation
    const binarySignatures = [
      [0xFF, 0xD8], [0x89, 0x50, 0x4E, 0x47], [0x47, 0x49, 0x46],
      [0x50, 0x4B, 0x03, 0x04], [0x25, 0x50, 0x44, 0x46]
    ];
    const skipExtensions = new Set([
      '.jpg', '.jpeg', '.png', '.gif', '.bmp', '.ico', '.webp',
      '.mp3', '.mp4', '.avi', '.mov', '.wav', '.flac',
      '.zip', '.tar', '.gz', '.rar', '.7z',
      '.pdf', '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx',
      '.sqlite', '.db', '.jar', '.class', '.pdb', '.lock', // Added .pdb, .lock
      '.icns', // Added .icns
    ]);
     const textExtensions = new Set([
      '.rs', '.ts', '.tsx', '.js', '.jsx', '.vue', '.svelte',
      '.html', '.css', '.scss', '.sass', '.less',
      '.json', '.toml', '.yaml', '.yml',
      '.md', '.markdown', '.txt', '.gitignore', '.taurignore', // Added gitignore/taurignore
      '.sh', '.bash', '.zsh', '.fish', '.bat', '.ps1', // Added bat/ps1
      '.c', '.cpp', '.h', '.hpp', '.cs', '.go', '.py', '.rb',
      '.sql', '.code-workspace', '.svg' // Added sql, code-workspace, svg
    ]);


    const fileStats = { read: 0, skipped: 0, binary: 0, tooLarge: 0, error: 0, emptyFile: 0 };

    for (const filePath of this.allFiles) {
        try {
            const ext = path.extname(filePath).toLowerCase();
            // Handle files with no extension - treat as text unless binary detected
            if (ext === '' && !this._isLikelyBinary(filePath, binarySignatures)) {
                 // Allow reading files with no extension if they seem like text
            } else if (skipExtensions.has(ext)) {
                fileStats.skipped++;
                continue;
            }


            const stats = fs.statSync(filePath);
            if (stats.size === 0) {
                fileStats.emptyFile++;
                continue;
            }

            const isKnownText = textExtensions.has(ext) || ext === ''; // Include no-extension files here
            const sizeLimit = isKnownText ? 5 * 1024 * 1024 : 1024 * 1024;
            if (stats.size > sizeLimit) {
                fileStats.tooLarge++;
                continue;
            }

            // Check binary status only if not explicitly known text or empty extension
            if (!isKnownText && this._isLikelyBinary(filePath, binarySignatures)) {
                fileStats.binary++;
                continue;
            }

            const content = fs.readFileSync(filePath, 'utf8');
            this.fileContents.set(filePath, content);
            this._indexFileKeywords(filePath, content);
            fileStats.read++;

        } catch (err) {
            // Handle potential permission errors more gracefully
            if (err.code === 'EACCES' || err.code === 'EPERM') {
                 console.warn(`Permission denied reading ${filePath}`);
                 fileStats.error++;
            } else {
                fileStats.error++;
                console.error(`Error reading ${filePath}:`, err.message);
            }
        }
    }

    console.log(`Read ${fileStats.read} files, ${fileStats.skipped + fileStats.binary + fileStats.tooLarge + fileStats.emptyFile + fileStats.error} skipped/error`);
    console.log(`  Skipped: ${fileStats.skipped} by extension, ${fileStats.binary} binary, ${fileStats.tooLarge} too large, ${fileStats.emptyFile} empty`);
    console.log(`  Errors: ${fileStats.error}`);
    return this.fileContents;
};


module.exports = FileSystemUtils;

File: src/utils/file_system_utils.rs
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};
use walkdir::WalkDir;
use regex::Regex;
use serde::{Serialize, Deserialize};

/// Project structure information
#[derive(Debug, Serialize, Deserialize, Default)]
pub struct ProjectStructure {
    pub directories: HashSet<String>,
    pub files_by_type: HashMap<String, Vec<String>>,
    pub files_by_dir: HashMap<String, Vec<String>>,
    pub dir_categories: DirCategories,
}

/// Directory categories for classification
#[derive(Debug, Serialize, Deserialize, Default)]
pub struct DirCategories {
    pub api: HashSet<String>,
    pub models: HashSet<String>,
    pub ui: HashSet<String>,
    pub tests: HashSet<String>,
    pub services: HashSet<String>,
}

/// File system utilities for project analysis
pub struct FileSystemUtils {
    base_dir: PathBuf,
    exclude_patterns: Vec<Regex>,
    all_files: Vec<PathBuf>,
    file_contents: HashMap<PathBuf, String>,
    project_structure: ProjectStructure,
    keyword_index: HashMap<String, Vec<PathBuf>>,
}

impl FileSystemUtils {
    /// Create a new FileSystemUtils instance
    pub fn new<P: AsRef<Path>>(base_dir: P, exclude_patterns: Vec<&str>) -> Self {
        let compiled_patterns = exclude_patterns
            .iter()
            .filter_map(|pattern| Regex::new(pattern).ok())
            .collect();

        Self {
            base_dir: base_dir.as_ref().to_path_buf(),
            exclude_patterns: compiled_patterns,
            all_files: Vec::new(),
            file_contents: HashMap::new(),
            project_structure: ProjectStructure::default(),
            keyword_index: HashMap::new(),
        }
    }

    /// Discover all relevant files in the base directory
    pub fn discover_files(&mut self) -> &Vec<PathBuf> {
        println!("Discovering files...");
        self.all_files = self.walk_dir(&self.base_dir);
        println!("Found {} files", self.all_files.len());
        &self.all_files
    }

    /// Walk directory and collect files
    fn walk_dir(&mut self, dir: &Path) -> Vec<PathBuf> {
        let mut file_list = Vec::new();

        for entry in WalkDir::new(dir)
            .follow_links(true)
            .into_iter()
            .filter_map(Result::ok)
        {
            let path = entry.path().to_path_buf();
            let relative_path = path
                .strip_prefix(&self.base_dir)
                .unwrap_or(&path)
                .to_string_lossy()
                .to_string();

            // Skip if matches any exclude pattern
            if self
                .exclude_patterns
                .iter()
                .any(|pattern| pattern.is_match(&relative_path))
            {
                continue;
            }

            if entry.file_type().is_dir() {
                // Track directory for project structure analysis
                if !relative_path.starts_with('.') {
                    // Avoid hidden dirs like .git
                    self.project_structure
                        .directories
                        .insert(relative_path.clone());
                    self.categorize_directory(&relative_path);
                }
            } else {
                // Process file
                let ext = path
                    .extension()
                    .unwrap_or_default()
                    .to_string_lossy()
                    .to_string();

                // Track by file type
                self.project_structure
                    .files_by_type
                    .entry(ext.clone())
                    .or_default()
                    .push(relative_path.clone());

                // Track by directory
                let dir_name = path
                    .parent()
                    .unwrap_or_else(|| Path::new(""))
                    .strip_prefix(&self.base_dir)
                    .unwrap_or_else(|_| Path::new(""))
                    .to_string_lossy()
                    .to_string();

                self.project_structure
                    .files_by_dir
                    .entry(dir_name)
                    .or_default()
                    .push(relative_path);

                file_list.push(path);
            }
        }

        file_list
    }

    /// Categorize directory based on its name
    fn categorize_directory(&mut self, relative_path: &str) {
        let categories = &mut self.project_structure.dir_categories;

        if relative_path.contains("api") || relative_path.contains("routes") {
            categories.api.insert(relative_path.to_string());
        } else if relative_path.contains("model") || relative_path.contains("entity") {
            categories.models.insert(relative_path.to_string());
        } else if relative_path.contains("component")
            || relative_path.contains("ui")
            || relative_path.contains("pages")
            || relative_path.contains("features")
        {
            categories.ui.insert(relative_path.to_string());
        } else if relative_path.contains("test") || relative_path.contains("spec") {
            categories.tests.insert(relative_path.to_string());
        } else if relative_path.contains("service")
            || relative_path.contains("util")
            || relative_path.contains("core")
            || relative_path.contains("shared")
        {
            categories.services.insert(relative_path.to_string());
        }
    }

    /// Read file contents, skipping binaries and large files
    pub fn read_file_contents(&mut self) -> &HashMap<PathBuf, String> {
        println!("Reading file contents with advanced processing...");

        let skip_extensions: HashSet<&str> = [
            ".jpg", ".jpeg", ".png", ".gif", ".bmp", ".ico", ".webp",
            ".mp3", ".mp4", ".avi", ".mov", ".wav", ".flac",
            ".zip", ".tar", ".gz", ".rar", ".7z",
            ".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
            ".sqlite", ".db", ".jar", ".class", ".pdb", ".lock",
            ".icns",
        ].iter().cloned().collect();

        let text_extensions: HashSet<&str> = [
            ".rs", ".ts", ".tsx", ".js", ".jsx", ".vue", ".svelte",
            ".html", ".css", ".scss", ".sass", ".less",
            ".json", ".toml", ".yaml", ".yml",
            ".md", ".markdown", ".txt", ".gitignore", ".taurignore",
            ".sh", ".bash", ".zsh", ".fish", ".bat", ".ps1",
            ".c", ".cpp", ".h", ".hpp", ".cs", ".go", ".py", ".rb",
            ".sql", ".code-workspace", ".svg",
        ].iter().cloned().collect();

        // Stats for tracking processed files
        let mut read = 0;
        let mut skipped = 0;
        let mut binary = 0;
        let mut too_large = 0;
        let mut empty_file = 0;
        let mut error = 0;

        for file_path in &self.all_files {
            if let Some(ext) = file_path.extension() {
                let ext_str = ext.to_string_lossy().to_lowercase();
                if skip_extensions.contains(ext_str.as_str()) {
                    skipped += 1;
                    continue;
                }

                match fs::metadata(file_path) {
                    Ok(stats) => {
                        if stats.len() == 0 {
                            empty_file += 1;
                            continue;
                        }

                        let is_known_text = text_extensions.contains(ext_str.as_str());
                        let size_limit = if is_known_text { 5 * 1024 * 1024 } else { 1024 * 1024 };
                        if stats.len() > size_limit {
                            too_large += 1;
                            continue;
                        }

                        // Skip binary check for known text files
                        if !is_known_text && self.is_likely_binary(file_path) {
                            binary += 1;
                            continue;
                        }

                        match fs::read_to_string(file_path) {
                            Ok(content) => {
                                self.file_contents.insert(file_path.clone(), content.clone());
                                self.index_file_keywords(file_path, &content);
                                read += 1;
                            }
                            Err(_) => {
                                error += 1;
                            }
                        }
                    }
                    Err(_) => {
                        error += 1;
                    }
                }
            }
        }

        println!(
            "Read {} files, {} skipped",
            read,
            skipped + binary + too_large + empty_file
        );
        println!(
            "  Skipped: {} by extension, {} binary, {} too large, {} empty",
            skipped, binary, too_large, empty_file
        );

        &self.file_contents
    }

    /// Check if a file is likely binary
    fn is_likely_binary(&self, file_path: &Path) -> bool {
        // Common binary file signatures
        let binary_signatures: Vec<&[u8]> = vec![
            &[0xFF, 0xD8],                 // JPEG
            &[0x89, 0x50, 0x4E, 0x47],     // PNG
            &[0x47, 0x49, 0x46],           // GIF
            &[0x50, 0x4B, 0x03, 0x04],     // ZIP/JAR/DOCX
            &[0x25, 0x50, 0x44, 0x46],     // PDF
        ];

        if let Ok(content) = fs::read(file_path) {
            if content.len() < 8 {
                return false;
            }

            // Check for binary signatures
            for signature in binary_signatures {
                if content.starts_with(signature) {
                    return true;
                }
            }

            // Check for high percentage of non-ASCII characters
            let non_ascii_count = content
                .iter()
                .take(1024)
                .filter(|&&b| b > 127 || (b < 32 && b != 9 && b != 10 && b != 13))
                .count();

            if content.len() > 0 {
                let non_ascii_ratio = non_ascii_count as f64 / content.len().min(1024) as f64;
                if non_ascii_ratio > 0.3 {
                    return true;
                }
            }
        }

        false
    }

    /// Index keywords in a file for faster searching
    fn index_file_keywords(&mut self, file_path: &Path, content: &str) {
        // List of important keywords to index
        let keywords = [
            // Models
            "struct", "enum", "trait", "impl", "class", "interface", "type", "model", "entity",
            // API
            "fn", "function", "route", "get", "post", "put", "delete", "api", "endpoint", "handler",
            "router", "controller",
            // UI
            "component", "function", "render", "return", "useState", "useEffect", "props", "view",
            "page", "layout",
            // Tests
            "test", "describe", "it", "expect", "assert", "mock",
            // General
            "import", "export", "module", "require", "use", "let", "const", "var", "pub", "async",
            "await",
        ];

        for keyword in keywords.iter() {
            if content.contains(keyword) {
                self.keyword_index
                    .entry(keyword.to_string())
                    .or_default()
                    .push(file_path.to_path_buf());
            }
        }
    }

    /// Find files matching a list of regex patterns
    pub fn find_files_by_patterns(&self, patterns: &[Regex]) -> Vec<PathBuf> {
        let mut matching_files = Vec::new();

        for file_path in &self.all_files {
            let path_str = file_path.to_string_lossy();
            if patterns.iter().any(|pattern| pattern.is_match(&path_str)) {
                matching_files.push(file_path.clone());
            }
        }

        matching_files
    }

    /// Filter files by pattern
    pub fn filter_files(&self, pattern: &Regex) -> Vec<PathBuf> {
        self.all_files
            .iter()
            .filter(|file| pattern.is_match(&file.to_string_lossy()))
            .cloned()
            .collect()
    }

    /// Get directory category (models, api, ui, etc.)
    pub fn get_dir_category(&self, dir_path: &str) -> String {
        let categories = &self.project_structure.dir_categories;
        
        if categories.models.contains(dir_path) {
            return "models".to_string();
        } else if categories.api.contains(dir_path) {
            return "api".to_string();
        } else if categories.ui.contains(dir_path) {
            return "ui".to_string();
        } else if categories.tests.contains(dir_path) {
            return "tests".to_string();
        } else if categories.services.contains(dir_path) {
            return "services".to_string();
        }
        
        "other".to_string()
    }

    /// Get file statistics
    pub fn get_file_stats(&self) -> FileStats {
        let all_files = &self.all_files;
        
        // Count files by type
        let js = all_files
            .iter()
            .filter(|file| {
                let ext = file.extension().unwrap_or_default().to_string_lossy();
                ext == "js" || ext == "jsx" || ext == "ts" || ext == "tsx"
            })
            .count();
            
        let rust = all_files
            .iter()
            .filter(|file| file.extension().unwrap_or_default() == "rs")
            .count();
            
        let other = all_files.len() - js - rust;
        
        FileStats {
            total: all_files.len(),
            js,
            rust,
            other,
        }
    }

    // --- Getters for accessed properties ---
    
    /// Get all discovered files
    pub fn get_all_files(&self) -> &Vec<PathBuf> {
        &self.all_files
    }

    /// Get map of file contents
    pub fn get_file_contents_map(&self) -> &HashMap<PathBuf, String> {
        &self.file_contents
    }

    /// Get project structure information
    pub fn get_project_structure(&self) -> &ProjectStructure {
        &self.project_structure
    }

    /// Get keyword index
    pub fn get_keyword_index(&self) -> &HashMap<String, Vec<PathBuf>> {
        &self.keyword_index
    }
}

/// File statistics
#[derive(Debug, Serialize, Deserialize)]
pub struct FileStats {
    pub total: usize,
    pub js: usize,
    pub rust: usize,
    pub other: usize,
}
